### PR TITLE
WIP: Add prototype AWS S3 Cache

### PR DESCRIPTION
**Pull request by nickrobinson251**

### CODE REVIEW

File: src/pudl/workspace/resource_cache.py [hunk 1]
```diff
@@ -135,6 +137,50 @@ def contains(self, resource: PudlResourceKey) -> bool:
...
+        """Deletes resource from the cache."""
+        return self._blob(resource).delete()
+
+    def contains(self, resource: PudlResourceKey) -> bool:
+        """Returns True if resource is present in the cache."""
+        try:
+            self._blob(resource).load()
+            return True
+        except:
+            return False
```
Comment (fchorney) [inline src/pudl/workspace/resource_cache.py:184]: you maybe want something more like this to be more robust in catching other errors.

```python
try:
    self._blob(resource).load()
except botocore.exceptions.ClientError as e:
    if e.response['Error']['Code'] == "404":
        # The object does not exist.
        return False
    else:
        # Something else has gone wrong.
        raise
else:
    # The object does exist.
    return True
```

File: src/pudl/workspace/resource_cache.py [hunk 2]
```diff
@@ -135,6 +137,50 @@ def contains(self, resource: PudlResourceKey) -> bool:
...
+        p = (self._path_prefix / resource.get_local_path()).as_posix().lstrip('/')
+        return self._bucket.Object(p)
+
+    def get(self, resource: PudlResourceKey) -> bytes:
+        """Retrieves value associated with given resource."""
+        return self._blob(resource).get()['Body'].read()
+
+    def add(self, resource: PudlResourceKey, value: bytes):
+        """Adds (or updates) resource to the cache with given value."""
+        # TODO return self._blob(resource).upload_from_string(value)
```
Comment (fchorney) [inline src/pudl/workspace/resource_cache.py]: Is this just a comment left over from porting the google cache code to s3?

File: src/pudl/workspace/datastore.py [hunk 3]
```diff
@@ -435,12 +447,13 @@ def _get_pudl_in(args: dict) -> Path:
 def _create_datastore(args: dict) -> Datastore:
     """Constructs datastore instance."""
     local_cache_path = None
-    if not args.populate_gcs_cache:
+    if not args.populate_gcs_cache and not args.populate_s3_cache:
```
Comment (fchorney) [inline src/pudl/workspace/datastore.py]: Either or is fine, just suggesting an alternative
```suggestion
    if not (args.populate_gcs_cache or args.populate_s3_cache):
```

File: src/pudl/dfc.py [hunk 4]
```diff
@@ -79,10 +79,10 @@ def get(self, name: str) -> pd.DataFrame:
     def store(self, name: str, data: pd.DataFrame):
         """Adds named dataframe to collection and stores its contents on disk."""
         filename = self._get_filename(name, self._instance_id)
-        if not filename.startswith("gs://"):
+        if not filename.startswith("gs://") or not filename.startswith("s3://"):
```
Comment (fchorney) [inline src/pudl/dfc.py]: Either or is fine, just suggesting an alternative
```suggestion
        if not (filename.startswith("gs://") and filename.startswith("s3://")):
```

File: src/pudl/workspace/resource_cache.py (hunk 1 above)
Comment (nickrobinson251) [reply src/pudl/workspace/resource_cache.py:184]: I was hoping you'd be able to suggest a more robust way of doing this 😂 

I knew we wouldn't want a plain `except` but wasn't sure where to start with something more robust - thanks!

File: src/pudl/dfc.py (hunk 4 above)
Comment (nickrobinson251) [reply src/pudl/dfc.py]: oh this logic should probably be 
```suggestion
        if not (filename.startswith("gs://") or filename.startswith("s3://")):
```
i suspect this code path didn't get hit in my local `pudl_datastore`/`pudl_etl` experiments with this branch. 

One for if/when we add tests and try to integrate this upstream :)

Comment (fchorney) [reply src/pudl/dfc.py]: ah yes good point!

**Pull request closed without merging**